rs/rosetta-api/icrc1/ledger/src/main.rs

```diff
@@ -135,18 +142,79 @@ fn pre_upgrade() {
         .expect("failed to write instructions consumed to stable memory");
 }
 
+#[cfg(feature = "upgrade-to-memory-manager")]
+#[pre_upgrade]
+fn pre_upgrade() {
+    #[cfg(feature = "canbench-rs")]
+    let _p = canbench_rs::bench_scope("pre_upgrade");
+
+    let start = ic_cdk::api::instruction_counter();
+    UPGRADES_MEMORY.with_borrow_mut(|bs| {
+        Access::with_ledger(|ledger| {
+            let writer = Writer::new(bs, 0);
+            let mut buffered_writer = BufferedWriter::new(8388608, writer);
+            ciborium::ser::into_writer(ledger, &mut buffered_writer)
+                .expect("Failed to write the Ledger state in stable memory");
+            let end = ic_cdk::api::instruction_counter();
+            let instructions_consumed = end - start;
+            let counter_bytes: [u8; 8] = instructions_consumed.to_le_bytes();
+            buffered_writer
+                .write_all(&counter_bytes)
+                .expect("failed to write instructions consumed to UPGRADES_MEMORY");
+        });
+    });
+}
+
 #[post_upgrade]
 fn post_upgrade(args: Option<LedgerArgument>) {
     #[cfg(feature = "canbench-rs")]
     let _p = canbench_rs::bench_scope("post_upgrade");
 
     let start = ic_cdk::api::instruction_counter();
     let mut stable_reader = StableReader::default();
-    LEDGER.with(|cell| {
-        *cell.borrow_mut() = Some(
-            ciborium::de::from_reader(&mut stable_reader).expect("failed to decode ledger state"),
-        );
-    });
+
+    let mut pre_upgrade_instructions_consumed = 0;
+
+    let old_des_result: Result<Ledger<Tokens>, _> = ciborium::de::from_reader(&mut stable_reader);
+
+    match old_des_result {
+        Ok(state) => {
+            LEDGER.with(|cell| {
+                *cell.borrow_mut() = Some(state);
+            });
+            let mut pre_upgrade_instructions_counter_bytes = [0u8; 8];
+            pre_upgrade_instructions_consumed =
+                match stable_reader.read_exact(&mut pre_upgrade_instructions_counter_bytes) {
+                    Ok(_) => u64::from_le_bytes(pre_upgrade_instructions_counter_bytes),
+                    Err(_) => {
+                        // If upgrading from a version that didn't write the instructions counter to stable memory
+                        0u64
+                    }
+                };
+        }
+        Err(_) => {
+            let state: Ledger<Tokens> = UPGRADES_MEMORY.with_borrow(|bs| {
+                let mut reader = Reader::new(bs, 0);
```

Comment:
Should we wrap the reader into a buffered reader (analogously to the writer)?

Reply:
Done!